### PR TITLE
[SPARK-48900] Add `reason` field for all internal calls for job/stage cancellation

### PR DESCRIPTION
**Pull request by mingkangli-db**
### What changes were proposed in this pull request?

The changes can be grouped into two categories:

- **Add `reason` field for all internal calls for job/stage cancellation**
  - Cancel because of exceptions:
    - org/apache/spark/sql/connect/execution/ExecuteThreadRunner.scala
    - org/apache/spark/sql/execution/exchange/BroadcastExchangeExec.scala
    - sql/hive-thriftserver/src/main/scala/org/apache/spark/sql/hive/thriftserver/SparkExecuteStatementOperation.scala
   - Cancel by user (Web UI):
      - org/apache/spark/ui/jobs/JobsTab.scala 
   - Cancel when streaming terminates/query ends:
     - org/apache/spark/sql/execution/streaming/MicroBatchExecution.scala
     - org/apache/spark/sql/execution/streaming/continuous/ContinuousExecution.scala
     
_(Developers familiar with these components: would appreciate it if you could provide suggestions for more helpful cancellation reason strings! :) )_

- **API Change for `JobWaiter`**
Currently, the `.cancel()` function in `JobWaiter` does not allow specifying a reason for cancellation. This limitation prevents us from reporting cancellation reasons in AQE, such as cleanup or canceling unnecessary jobs after query replanning. To address this, we should add an `reason: String` parameter to all relevant APIs along the chain.

image

### Why are the changes needed?

Today it is difficult to determine why a job, stage, or job group was canceled. We should leverage existing Spark functionality to provide a reason string explaining the cancellation cause, and should add new APIs to let us provide this reason when canceling job groups. For more context, please read [this JIRA ticket](https://issues.apache.org/jira/browse/SPARK-48900).

This feature can be implemented in two PRs:
1. [Modify the current SparkContext and its downstream APIs to add the reason string, such as cancelJobGroup and cancelJobsWithTag](https://github.com/apache/spark/pull/47361)
2. Add reasons for all internal calls to these methods.

**Note: This is the second of the two PRs to implement this new feature**

### Does this PR introduce _any_ user-facing change?

It adds reasons for jobs and stages cancelled internally, providing users with clearer explanations for cancellations by Spark, such as exceptions, end of streaming, AQE query replanning, etc.

### How was this patch tested?

Tests for the API changes on the `JobWaiter` chain are added in `JobCancellationSuite.scala`

### Was this patch authored or co-authored using generative AI tooling?

No.

### CODE REVIEW

File: sql/core/src/test/scala/org/apache/spark/sql/execution/BroadcastExchangeSuite.scala [hunk 1]
```diff
@@ -82,7 +83,7 @@ class BroadcastExchangeSuite extends SparkPlanTest
       val events = jobEvents.toArray
       val hasStart = events(0).isInstanceOf[SparkListenerJobStart]
       val hasCancelled = events(1).asInstanceOf[SparkListenerJobEnd].jobResult
-        .asInstanceOf[JobFailed].exception.getMessage.contains("cancelled job tag")
+        .asInstanceOf[JobFailed].exception.getMessage.contains("Could not execute broadcast")
```
Comment (mingkangli-db) [inline sql/core/src/test/scala/org/apache/spark/sql/execution/BroadcastExchangeSuite.scala:86]: This is changed because now on timeout, `BroadcastExchangeExec` will now print out the exception, not a generic statement such as "cancelled job tag". See [this line](https://github.com/apache/spark/blob/0444ba32156e424ee6d902c3f4a4c44a07993273/sql/core/src/main/scala/org/apache/spark/sql/execution/exchange/BroadcastExchangeExec.scala#L241)

File: sql/core/src/main/scala/org/apache/spark/sql/execution/exchange/ShuffleExchangeExec.scala [hunk 2]
```diff
@@ -83,10 +83,23 @@ trait ShuffleExchangeLike extends Exchange {
 
   protected def mapOutputStatisticsFuture: Future[MapOutputStatistics]
 
+  /**
+   * Cancels the shuffle job with a reason.
+   */
+  final def cancelShuffleJob(reason: String): Unit = {
```
Comment (JoshRosen) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/exchange/ShuffleExchangeExec.scala:89]: I'm not a fan of the copy-paste code duplication between the "with reason" and "without reason" overloads and think that we should instead refactor the common code to a `private final def doCancelShuffleJob(reason: Option(reason))`. Within that method, we can do something like

```scala
reason match {
  case Some(reason) => action.cancel(reason()
  case None => action.cancel()
}
```

or similar.

File: sql/core/src/main/scala/org/apache/spark/sql/execution/exchange/BroadcastExchangeExec.scala [hunk 3]
```diff
@@ -73,6 +73,16 @@ trait BroadcastExchangeLike extends Exchange {
 
   protected def completionFuture: scala.concurrent.Future[broadcast.Broadcast[Any]]
 
+  /**
+   * Cancels broadcast job with a reason.
+   */
+  final def cancelBroadcastJob(reason: String): Unit = {
```
Comment (JoshRosen) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/exchange/BroadcastExchangeExec.scala:79]: Likewise, any way that we can factor out the common duplicated code here as well?

File: project/MimaExcludes.scala [hunk 4]
```diff
@@ -100,7 +100,13 @@ object MimaExcludes {
     ProblemFilters.exclude[DirectMissingMethodProblem]("org.apache.spark.sql.SQLContext#implicits._sqlContext"),
     ProblemFilters.exclude[DirectMissingMethodProblem]("org.apache.spark.sql.SQLImplicits._sqlContext"),
     ProblemFilters.exclude[ReversedMissingMethodProblem]("org.apache.spark.sql.SQLImplicits.session"),
-    ProblemFilters.exclude[DirectMissingMethodProblem]("org.apache.spark.sql.SparkSession#implicits._sqlContext")
+    ProblemFilters.exclude[DirectMissingMethodProblem]("org.apache.spark.sql.SparkSession#implicits._sqlContext"),
+    // SPARK-48900: Add `reason` string to all job / stage / job group cancellation calls
+    ProblemFilters.exclude[ReversedMissingMethodProblem]("org.apache.spark.scheduler.JobWaiter#cancel"),
```
Comment (JoshRosen) [inline project/MimaExcludes.scala:105]: `ReversedMissingMethodProblem` is okay to suppress here, since that one is warning that we have an interface abstract / class which has gained an abstract method such that old implementations might hit an error when linked against the updated interface / abstract class. I don't think we intend to support _user-defined implementations of FutureAction_, so it's okay if Spark 4.0 imposes new requirements on this.

But `DirectMissingMethodProblem` is **not** okay to ignore: that's saying that we're _removing_  a method which used to be present in older versions, which will call callers to fail. We should add back a zero-argument overload so that calls of the old signature will continue to work.

File: core/src/main/scala/org/apache/spark/FutureAction.scala [hunk 5]
```diff
@@ -39,9 +39,9 @@ trait FutureAction[T] extends Future[T] {
   // documentation (with reference to the word "action").
 
   /**
-   * Cancels the execution of this action.
+   * Cancels the execution of this action with an optional reason.
    */
-  def cancel(): Unit
+  def cancel(reason: Option[String]): Unit
```
Comment (JoshRosen) [inline core/src/main/scala/org/apache/spark/FutureAction.scala:44]: As I also commented on MimaExcludes, it is okay to _add_ a new overload which accepts an optional reason, but we should not _remove_ the old overload.

Instead, I think we should keep the zero-argument `def cancel(): Unit` and provide a default implementation which simply calls `cancel(None)`.

We don't need to override the zero-argument `cancel()` in subclasses; those can remain unchanged.

If we do that then we won't have to add in explicit `cancel(reason = None)` in a bunch of places, which will help to shrink the size of this PR.

File: connect/server/src/main/scala/org/apache/spark/sql/connect/execution/ExecuteThreadRunner.scala [hunk 6]
```diff
@@ -115,7 +116,8 @@ private[connect] class ExecuteThreadRunner(executeHolder: ExecuteHolder) extends
         case e: Throwable =>
           logDebug(s"Exception in execute: $e")
           // Always cancel all remaining execution after error.
-          executeHolder.sessionHolder.session.sparkContext.cancelJobsWithTag(executeHolder.jobTag)
+          executeHolder.sessionHolder.session.sparkContext.cancelJobsWithTag(executeHolder.jobTag,
+            "Exception in execute:" + ExceptionUtils.getStackTrace(e))
```
Comment (JoshRosen) [inline connect/server/src/main/scala/org/apache/spark/sql/connect/execution/ExecuteThreadRunner.scala:120]: Huh, looks like `ExceptionUtils.getStackTrace` is approximately equivalent to Spark's own `Utils.exceptionString` but it seems that we're not very consistent about using one vs. the other.

I think you should add an extra space after `execute:` so this concatenates properly.

Comment (mingkangli-db) [reply connect/server/src/main/scala/org/apache/spark/sql/connect/execution/ExecuteThreadRunner.scala]: Sounds good, I also switched to using `Utils.exceptionString` because it seems to be more commonly used than the other in the codebase.

File: core/src/main/scala/org/apache/spark/FutureAction.scala [hunk 7]
```diff
@@ -39,9 +39,14 @@ trait FutureAction[T] extends Future[T] {
...
+
+  /**
+   * Cancels the execution of this action with an optional reason.
```
Comment (jiangxb1987) [inline core/src/main/scala/org/apache/spark/FutureAction.scala:47]: nit: update the comment.

File: core/src/test/scala/org/apache/spark/JobCancellationSuite.scala [hunk 8]
```diff
@@ -603,7 +603,8 @@ class JobCancellationSuite extends SparkFunSuite with Matchers with BeforeAndAft
     // Kill one of the action.
     Future {
       sem1.acquire()
-      f1.cancel()
+      val reason = "custom cancel reason"
+      f1.cancel(Option(reason))
```
Comment (jiangxb1987) [inline core/src/test/scala/org/apache/spark/JobCancellationSuite.scala:607]: Where do we verify the reason? If we don't need to verify, then maybe revert the change here?

File: sql/core/src/main/scala/org/apache/spark/sql/execution/adaptive/AdaptiveSparkPlanExec.scala [hunk 9]
```diff
@@ -827,7 +827,7 @@ case class AdaptiveSparkPlanExec(
       // so we should avoid calling cancel on it to re-trigger the failure again.
       case s: ExchangeQueryStageExec if !earlyFailedStage.contains(s.id) =>
         try {
-          s.cancel()
+          s.cancel(Option("cancelled by cleanup in AdaptiveSparkPlanExec"))
```
Comment (jiangxb1987) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/adaptive/AdaptiveSparkPlanExec.scala:830]: nit: Option("...") => Some("...")

File: sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/continuous/ContinuousExecution.scala [hunk 10]
```diff
@@ -334,7 +334,7 @@ class ContinuousExecution(
           epochUpdateThread.interrupt()
           epochUpdateThread.join()
           // The following line must be the last line because it may fail if SparkContext is stopped
-          sparkSession.sparkContext.cancelJobGroup(runId.toString)
+          sparkSession.sparkContext.cancelJobGroup(runId.toString, "continuous execution finished")
```
Comment (jiangxb1987) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/continuous/ContinuousExecution.scala:337]: Is it possible to include something similar to a query id within the reason?

File: core/src/test/scala/org/apache/spark/JobCancellationSuite.scala (hunk 8 above)
Comment (mingkangli-db) [reply core/src/test/scala/org/apache/spark/JobCancellationSuite.scala:607]: Sounds good, I reverted the change for this test and added another short test that verifies the reason for `FutureAction.cancel()`

File: sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/continuous/ContinuousExecution.scala (hunk 10 above)
Comment (mingkangli-db) [reply sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/continuous/ContinuousExecution.scala:337]: Modified the reason to include `runId`.

File: connect/server/src/main/scala/org/apache/spark/sql/connect/execution/ExecuteThreadRunner.scala [hunk 11]
```diff
@@ -115,7 +115,8 @@ private[connect] class ExecuteThreadRunner(executeHolder: ExecuteHolder) extends
         case e: Throwable =>
           logDebug(s"Exception in execute: $e")
           // Always cancel all remaining execution after error.
-          executeHolder.sessionHolder.session.sparkContext.cancelJobsWithTag(executeHolder.jobTag)
+          executeHolder.sessionHolder.session.sparkContext.cancelJobsWithTag(executeHolder.jobTag,
+            "Exception in execute: " + Utils.exceptionString(e))
```
Comment (cloud-fan) [inline connect/server/src/main/scala/org/apache/spark/sql/connect/execution/ExecuteThreadRunner.scala:119]: I'm a bit worried about putting the full exception string in the reason field, as it can be very long. How about `A job with the same tag '$tagName' has failed`

File: core/src/main/scala/org/apache/spark/FutureAction.scala [hunk 12]
```diff
@@ -39,9 +39,11 @@ trait FutureAction[T] extends Future[T] {
   // documentation (with reference to the word "action").
 
   /**
-   * Cancels the execution of this action.
+   * Cancels the execution of this action with an optional reason.
    */
-  def cancel(): Unit
+  def cancel(reason: Option[String]): Unit
+
+  def cancel(): Unit = cancel(None)
```
Comment (cloud-fan) [inline core/src/main/scala/org/apache/spark/FutureAction.scala:46]: let's add back the API doc

File: core/src/main/scala/org/apache/spark/scheduler/JobWaiter.scala [hunk 13]
```diff
@@ -44,15 +44,18 @@ private[spark] class JobWaiter[T](
...
+   * cancellation itself is handled asynchronously. After the low level scheduler cancels
+   * all the tasks belonging to this job, it will fail this job with a SparkException.
    */
-  def cancel(): Unit = {
-    dagScheduler.cancelJob(jobId, None)
+  def cancel(reason: Option[String]): Unit = {
+    dagScheduler.cancelJob(jobId, reason)
   }
 
+  def cancel(): Unit = cancel(None)
```
Comment (cloud-fan) [inline core/src/main/scala/org/apache/spark/scheduler/JobWaiter.scala:57]: ditto, add back its API doc

File: sql/core/src/main/scala/org/apache/spark/sql/execution/adaptive/AdaptiveSparkPlanExec.scala [hunk 14]
```diff
@@ -827,7 +827,7 @@ case class AdaptiveSparkPlanExec(
       // so we should avoid calling cancel on it to re-trigger the failure again.
       case s: ExchangeQueryStageExec if !earlyFailedStage.contains(s.id) =>
         try {
-          s.cancel()
+          s.cancel(Some("cancelled by cleanup in AdaptiveSparkPlanExec"))
```
Comment (cloud-fan) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/adaptive/AdaptiveSparkPlanExec.scala:830]:
```suggestion
          s.cancel(Some("The corresponding SQL query has failed."))
```

File: sql/core/src/main/scala/org/apache/spark/sql/execution/adaptive/QueryStageExec.scala [hunk 15]
```diff
@@ -154,14 +154,14 @@ abstract class QueryStageExec extends LeafExecNode {
 abstract class ExchangeQueryStageExec extends QueryStageExec {
 
   /**
-   * Cancel the stage materialization if in progress; otherwise do nothing.
+   * Cancel the stage materialization if in progress with an optional reason; otherwise do nothing.
    */
-  final def cancel(): Unit = {
+  final def cancel(reason: Option[String]): Unit = {
```
Comment (cloud-fan) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/adaptive/QueryStageExec.scala:159]: does any caller pass `None`?

File: sql/core/src/main/scala/org/apache/spark/sql/execution/exchange/BroadcastExchangeExec.scala [hunk 16]
```diff
@@ -227,7 +230,8 @@ case class BroadcastExchangeExec(
       case ex: TimeoutException =>
         logError(log"Could not execute broadcast in ${MDC(TIMEOUT, timeout)} secs.", ex)
         if (!relationFuture.isDone) {
-          sparkContext.cancelJobsWithTag(jobTag)
+          sparkContext.cancelJobsWithTag(jobTag,
+            s"Could not execute broadcast in ${timeout} secs: " + Utils.exceptionString(ex))
```
Comment (cloud-fan) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/exchange/BroadcastExchangeExec.scala:234]: `The corresponding broadcast query has failed.`

File: sql/core/src/test/scala/org/apache/spark/sql/execution/BroadcastExchangeSuite.scala [hunk 17]
```diff
@@ -34,7 +34,8 @@ import org.apache.spark.tags.ExtendedSQLTest
 @ExtendedSQLTest
 class BroadcastExchangeSuite extends SparkPlanTest
   with SharedSparkSession
-  with AdaptiveSparkPlanHelper {
+  with AdaptiveSparkPlanHelper
+  {
```
Comment (cloud-fan) [inline sql/core/src/test/scala/org/apache/spark/sql/execution/BroadcastExchangeSuite.scala:38]: unnecessary change

File: sql/hive-thriftserver/src/main/scala/org/apache/spark/sql/hive/thriftserver/SparkExecuteStatementOperation.scala [hunk 18]
```diff
@@ -249,10 +249,11 @@ private[hive] class SparkExecuteStatementOperation(
       case e: Throwable =>
         // When cancel() or close() is called very quickly after the query is started,
         // then they may both call cleanup() before Spark Jobs are started. But before background
-        // task interrupted, it may have start some spark job, so we need to cancel again to
+        // task interrupted, it may have started some spark job, so we need to cancel again to
         // make sure job was cancelled when background thread was interrupted
         if (statementId != null) {
-          session.sparkContext.cancelJobGroup(statementId)
+          session.sparkContext.cancelJobGroup(statementId,
+            s"failed to execute statement ${redactedStatement}: " + SparkUtils.exceptionString(e))
```
Comment (cloud-fan) [inline sql/hive-thriftserver/src/main/scala/org/apache/spark/sql/hive/thriftserver/SparkExecuteStatementOperation.scala:256]: `The corresponding Thriftserver query has failed.`

File: sql/core/src/main/scala/org/apache/spark/sql/execution/adaptive/QueryStageExec.scala (hunk 15 above)
Comment (mingkangli-db) [reply sql/core/src/main/scala/org/apache/spark/sql/execution/adaptive/QueryStageExec.scala:159]: I don't think so, are you suggesting we can change the method signature to `reason: String` then?

File: sql/core/src/main/scala/org/apache/spark/sql/execution/adaptive/QueryStageExec.scala [hunk 19]
```diff
@@ -154,14 +154,14 @@ abstract class QueryStageExec extends LeafExecNode {
...
    */
-  final def cancel(): Unit = {
+  final def cancel(reason: String): Unit = {
     logDebug(s"Cancel query stage: $name")
-    doCancel()
+    doCancel(Option(reason))
   }
 
-  protected def doCancel(): Unit
+  protected def doCancel(reason: Option[String]): Unit
```
Comment (cloud-fan) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/adaptive/QueryStageExec.scala:164]: where do we call `doCancel` without a reason?

**Review by cloud-fan — APPROVED** [review]

File: sql/core/src/main/scala/org/apache/spark/sql/execution/adaptive/QueryStageExec.scala (hunk 19 above)
Comment (mingkangli-db) [reply sql/core/src/main/scala/org/apache/spark/sql/execution/adaptive/QueryStageExec.scala:164]: I don't think we ever call it without a reason. Changed the type from `Option[String]` to `String`.

**Comment by cloud-fan** [comment]
thanks, merging to master!
